README.md
node-iso11649
=============

[![CircleCI](https://circleci.com/gh/nruotsal/node-iso11649/tree/master.svg?style=svg)](https://circleci.com/gh/nruotsal/node-iso11649/tree/master)
[![Known Vulnerabilities](https://snyk.io/test/github/nruotsal/node-iso11649/badge.svg)](https://snyk.io/test/github/nruotsal/node-iso11649)

ISO 11649:2009 RF creditor reference library for node

* No dependencies
* 4 KB (1.2 KB gzipped)


## Installation

with NPM
```sh
npm install node-iso11649 --save
```

with Yarn
```sh
yarn add node-iso11649
```


## Usage

### Generating RF creditor reference

RF creditor reference can be generated from existing reference.

Existing reference characteristics:
 * Contain only numbers 0-9 and/or characters A-Z (example AB2G5 => RF68AB2G5).
 * Max length 21 characters.
 * Not case sensitive (example aB2g5 => RF68AB2G5).
 * Can be string with spaces (example '12345 12345' => RF451234512345).

```js
import { generate } from 'node-iso11649'

console.log(generate('12345 12345'))
// => RF451234512345
```

RF creditor reference can be generated also without existing reference.
In that case epoch timestamp increased with random number and finnish checksum at the end is used.

**NOTE**: While this provides some level of randomness, uniqueness is not guaranteed.
In production, it is best to use cryptographically secure random numbers at least as a part of the reference.

Example result:

```js
import { generate } from 'node-iso11649'

console.log(generate())
// => RF4714508655422864
```

As creditor references are commonly displayed in groups of 4 characters
you can use the optional `pretty` flag to format the returned value.

```js
import { generate } from 'node-iso11649'

console.log(generate('12345 12345', {pretty: true}))
// => RF45 1234 5123 45
```

### Validating RF creditor reference

Valid RF creditor reference characteristics:
 * Must start with characters RF.
 * Must contain two checksum numbers in indexes 3 and 4.
 * Reference part must follow rules described in 'Existing reference characteristics' section.

```js
import { validate } from 'node-iso11649'

console.log(validate('RF4714508655422864'))
// => true
```

```js
import { validate } from 'node-iso11649'

console.log(validate('RF00TEST'))
// => false
```

### Parsing RF creditor reference

The reference part can be extracted from the RF creditor reference.
This validates the entered RF creditor reference according to the
'Validating RF creditor reference'. It returns the reference string
if valid and null if invalid.

```js
import { parse } from 'node-iso11649'

console.log(parse('RF47 1450 8655 4228 64'))
// => 14508655422864
```

```js
import { parse } from 'node-iso11649'

console.log(parse('RF00TEST'))
// => null
```


## Tests

Run [ESLint](https://eslint.org/) style checker.

```sh
npm run lint
````

Run unit tests

```sh
npm test
```


CHANGELOG.md
# Changelog
This project uses [Semantic Versioning](http://semver.org/)

## 3.0.0 (2023-18-11)
**BREAKING**: Change generate function signature to have separate parameters for reference and options

- Improve randomness and uniqueness of RF references generated without existing reference
- Update dependencies
- Replace deprecated `substr` with `slice`
- Rollup type declaration file
- Improve documentation

## 2.1.2 (2020-01-10)
- [Fix modulo calculation](https://github.com/nruotsal/node-iso11649/pull/180) by [Tol1](https://github.com/Tol1)

## 2.1.1 (2020-28-04)
- [Fix types](https://github.com/nruotsal/node-iso11649/pull/170) by [davidknezic](https://github.com/davidknezic)

## 2.1.0 (2020-30-03)
- [Add parse function](https://github.com/nruotsal/node-iso11649/pull/150) by [davidknezic](https://github.com/davidknezic)
- [Add pretty print](https://github.com/nruotsal/node-iso11649/pull/149) by [davidknezic](https://github.com/davidknezic)

## 2.0.2 (2020-28-03)
- Remove unnecessary dependency

## 2.0.1 (2020-28-03)
- Fix main entrypoint

## 2.0.0 (2020-28-03)
- Rewrite with TypeScript

## 1.2.2 (2019-15-12)
- Update dependencies
- Build with latest Babel

## 1.2.1 (2018-16-08)
- Include only dist folder in npm package

## 1.2.0 (2018-16-08)
- TypeScript types
- Update dependencies
- Changelog in separate file

## 1.1.1 (2018-03-01)
- Bundle lib using rollup

## 1.1.0 (2018-02-27)
- Update dependencies
- Refactor code for easier maintenance

## 1.0.1 (2015-12-23)
- Handle number references as string
- Update README

## 1.0.0 (2015-12-23)
- Initial release


src/common.ts
const REFERENCE_FORMAT: RegExp = /^RF[0-9]{2}[0-9A-Z]+$/

type CharTable = Record<string, number>

/* eslint-disable object-property-newline */
const charTable: CharTable = {
  A: 10, B: 11, C: 12, D: 13, E: 14, F: 15, G: 16,
  H: 17, I: 18, J: 19, K: 20, L: 21, M: 22, N: 23,
  O: 24, P: 25, Q: 26, R: 27, S: 28, T: 29, U: 30,
  V: 31, W: 32, X: 33, Y: 34, Z: 35
}
/* eslint-disable object-property-newline */

export const normalizeReference = (reference: string): string =>
  (`${reference}`).replace(/ /g, '').toUpperCase()

export const substituteCharWithNumber = (char: string): string | number =>
  (Number.isNaN(Number(char)) ? charTable[char] ?? '' : char)

export const modulo97 = (dividend: string): number => {
  const chunks = dividend.match(/.{1,7}/g)
  return chunks !== null
    ? chunks.reduce((prev: number, curr: string) => parseInt(`${prev}${curr}`) % 97, 0)
    : -1
}

export const moveRfToEnd = (reference: string): string[] =>
  (reference.slice(4) + reference.slice(0, 4)).split('')

export const isValidChecksum = (reference: string): boolean => {
  const preResult = moveRfToEnd(reference).map(substituteCharWithNumber).join('')
  return modulo97(preResult) === 1
}

export const isValidFormat = (reference: string): boolean =>
  reference.match(REFERENCE_FORMAT) !== null

export const isValidChecksumRange = (reference: string): boolean => {
  const checkSum = Number(reference.slice(2, 4))
  return checkSum >= 2 && checkSum <= 98
}

export const isValid = (reference: string): boolean =>
  reference.length <= 25 &&
    isValidFormat(reference) &&
    isValidChecksumRange(reference) &&
    isValidChecksum(reference)
